pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -456,32 +452,28 @@ func (nim *nodeInfoManager) removeCSINodeInfo(csiDriverName string) error {
 
 		nodeInfoClient := csiKubeClient.CsiV1alpha1().CSINodeInfos()
 		nodeInfo, err := nodeInfoClient.Get(string(nim.nodeName), metav1.GetOptions{})
-		if nodeInfo == nil || errors.IsNotFound(err) {
-			// do nothing
-			return nil
-		}
 		if err != nil {
 			return err // do not wrap error
 		}
 
-		// Remove matching driver from driver list
+		hasModified := false
 		var newDriverInfos []csiv1alpha1.CSIDriverInfo
 		for _, driverInfo := range nodeInfo.CSIDrivers {
-			if driverInfo.Driver != csiDriverName {
-				newDriverInfos = append(newDriverInfos, driverInfo)
+			if driverInfo.Spec.Driver == csiDriverName {
+				// Uninstall the driver if we find it
+				driverInfo.Status.Available = false
+				hasModified = true
 			}
+			newDriverInfos = append(newDriverInfos, driverInfo)
 		}
 
-		if len(newDriverInfos) == len(nodeInfo.CSIDrivers) {
+		nodeInfo.CSIDrivers = newDriverInfos
+
+		if !hasModified {
```

Comment:
It's feels potentially problematic that we are adding and removing drivers one by one imperatively rather then periodically applying this list of drivers that the node currently knows to be installed. What if the nodeinfomanager is working off a cached nodeInfo here? It's possible that the apiserver served the object from cache. Would we ever retry the uninstall or would we lose track of the driver?

Reply:
I would agree, I believe this issue is tracked here: https://github.com/kubernetes/kubernetes/issues/67972 under the third point. I think for now this issue is out of the scope of this API Change